ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -277,34 +351,424 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
+		 * allocation comments for more details.
+		 */
+		n2_buf_size += n2_arr_size * (n1_buf_size / n2_buf_size - 1);
+		calc_size = n1_buf_size;
+	} else if (n1_buf_size < n2_buf_size) {
+		/*
+		 * If arr_size is small, size adjustment will be performed. See the memory
+		 * allocation comments for more details.
+		 */
+		n1_buf_size += n1_arr_size * (n2_buf_size / n1_buf_size - 1);
+		calc_size = n2_buf_size;
+	} else {
+		calc_size = n1_buf_size;
+	}
+
+	/*
+	 * How to calculate memory size
+	 *
+	 * n1_uint, n2_uint:
+	 * Use arr_size. The sum of these is prod_arr_size.
+	 *
+	 * n1_buf, n2_buf:
+	 * These are the necessary buffers for the divide and conquer "mid".
+	 * Assuming the computation length is M, the first split uses a buffer of M/2, and the second
+	 * split uses a buffer of M/4.... The required buffer size increases by half until the size
+	 * reaches 2. In other words, it is the sum of a geometric progression of 2 with a geometric
+	 * ratio of 2 from 2 to N-1, where the calculation length is N.
+	 * This sum can be calculated using the following formula, where the first term is a and the
+	 * geometric ratio is r: a(r^(N-1) - 1)(r - 1)
+	 * Here, a and r are both 2, so this formula becomes: 2(2^(N-1) - 1)(2 - 1) = 2^N - 2
+	 * In other words, the calculation length minus 2 is the required buffer size.
+	 *
+	 * However, if n_arr_size is very small compared to the calculation length, may not need a buffer
+	 * half of the calculation length when dividing. If n_arr_size is smaller than half of the calculation
+	 * length, the buffer size required during the division is equal to n_arr_size.
+	 * Therefore, the buffer size required in this case is the number of divisions until n_arr_size is
+	 * larger than half of the calculation length * n_arr_size + the sum of the calculations of the geometric
+	 * progression after that calculation length.
+	 * e.g. 62size(n1) * 7size(n2) => 64size * 8size (calc_size = 64)
+	 * divide and conquer 1: 32size * 8size (n2 requires 8 buf size)
+	 * divide and conquer 2: 16size * 8size (n2 requires 8 buf size)
+	 * divide and conquer 3: 8size * 8size (n2 requires 8 buf size)
+	 * divide and conquer 4: 4size * 4size (n2 requires 4 buf size)
+	 * divide and conquer 5: 2size * 2size (n2 requires 2 buf size)
+	 * So, the required buffer size is (8 * 3) + (2^3 - 2) = 30.
+	 *
+	 * prod_uint:
+	 * Share the results and the buffers used in intermediate calculations. The result is prod_arr_size.
+	 * The buffer increases by half like n_buf, but when calc_size is 2, the required buffer size is 4.
+	 * In other words, they are almost the same geometric progression, but the first term is 4.
+	 * a(r^(N-1) - 1)(r - 1): 4(2^(N-1) - 1)(2 - 1) = 2^N - 2 = 2(2^N - 2)
+	 * Therefore, the required size is prod_arr_size + calc_size * 2 - 4.
+	 */
+	n1_buf_size -= 2;
+	n2_buf_size -= 2;
+
+	BC_UINT_T *buf = safe_emalloc(prod_arr_size * 2 + calc_size * 2 - 4 + n1_buf_size + n2_buf_size, sizeof(BC_UINT_T), 0);
```

Comment:
Can this sum overflow?

Reply:
You are right. For very large numbers it will overflow...